From d24c441bb7338b178b3558b6f70cc82fe82b9088 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Mon, 24 Jul 2017 10:48:29 +0800
Subject: [PATCH 2214/5242] MLK-16075-19 gpu: imx: dpu: common: Add fetchecos
 support in dpu plane group

commit  0b6110d1973a781bafb064e58572988cde628239 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds fetchecos support in dpu plane group.
We currently supports fetcheco0 and fetcheco1.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-common.c |   10 ++++++++++
 include/video/dpu.h              |    2 ++
 2 files changed, 12 insertions(+)

diff --git a/drivers/gpu/imx/dpu/dpu-common.c b/drivers/gpu/imx/dpu/dpu-common.c
index 5ab02e2..15f14ef 100644
--- a/drivers/gpu/imx/dpu/dpu-common.c
+++ b/drivers/gpu/imx/dpu/dpu-common.c
@@ -866,6 +866,12 @@ static int dpu_get_plane_resource(struct dpu_soc *dpu,
 		if (IS_ERR(res->fd[i]))
 			return PTR_ERR(res->fd[i]);
 	}
+	for (i = 0; i < ARRAY_SIZE(res->fe); i++) {
+		res->fe[i] = dpu_fe_get(dpu, i);
+		if (IS_ERR(res->fe[i]))
+			return PTR_ERR(res->fe[i]);
+		grp->hw_plane_fetcheco_num = ARRAY_SIZE(res->fe);
+	}
 	/* HScaler could be shared with capture. */
 	if (display_plane_video_proc) {
 		for (i = 0; i < ARRAY_SIZE(res->hs); i++) {
@@ -912,6 +918,10 @@ static void dpu_put_plane_resource(struct dpu_plane_res *res)
 		if (!IS_ERR_OR_NULL(res->fd[i]))
 			dpu_fd_put(res->fd[i]);
 	}
+	for (i = 0; i < ARRAY_SIZE(res->fe); i++) {
+		if (!IS_ERR_OR_NULL(res->fe[i]))
+			dpu_fe_put(res->fe[i]);
+	}
 	for (i = 0; i < ARRAY_SIZE(res->hs); i++) {
 		if (!IS_ERR_OR_NULL(res->hs[i]))
 			dpu_hs_put(res->hs[i]);
diff --git a/include/video/dpu.h b/include/video/dpu.h
index d099e4f..bd2582e 100644
--- a/include/video/dpu.h
+++ b/include/video/dpu.h
@@ -658,6 +658,7 @@ struct dpu_plane_res {
 	struct dpu_constframe	*cf[2];
 	struct dpu_extdst	*ed[2];
 	struct dpu_fetchdecode	*fd[MAX_FD_NUM];
+	struct dpu_fetcheco	*fe[2];
 	struct dpu_framegen	*fg;
 	struct dpu_hscaler	*hs[2];
 	struct dpu_layerblend	*lb[MAX_LB_NUM];
@@ -673,6 +674,7 @@ struct dpu_plane_grp {
 	struct list_head	list;
 	struct mutex		lock;
 	unsigned int		hw_plane_num;
+	unsigned int		hw_plane_fetcheco_num;
 	unsigned int		hw_plane_hscaler_num;
 	unsigned int		hw_plane_vscaler_num;
 	unsigned int		id;
-- 
1.7.9.5

